refactor: audio implementation

Landeplage opened the pull request:
# What?
Refactored the implementation of audio, using Scriptable Objects for all audio events.

# Why?
To prevent issues with git merging while working further on the audio implementation, and for scalability. Also, because scriptable objects are just plain neat.

Landeplage committed:
refactor: Edit AudioEvents into Scriptable Objects, Add events for common HUD-elements

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation

Landeplage committed:
refactor: Added more AudioEvents and refactored ButtonAudioHandler

Landeplage committed:
refactor: ToggleAudioHandler

Also fixed an issue where cooldown in AudioEvent was applied at start of game.

Landeplage committed:
refactor: Initial reimplementation of audio for AvatarEditorHUD and FriendsHUD

Landeplage committed:
refactor: Toggle audio handlers

Landeplage committed:
refactor: Audio implementation of AvatarEditorHUD

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation

Landeplage committed:
refactor: Refactored audio implementation of various audio handler classes

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation

Landeplage committed:
refactor: Re-implemented camera sound and various dialog appear/close sounds

Landeplage committed:
refactor: Re-implemented all common HUD audio events

Landeplage committed:
refactor: Saved scene

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation

Landeplage committed:
refactor: Re-implemented map parcel cursor hover sound

Landeplage committed:
refactor: Re-implemented explore HUD, fixed bug with fade in sound

Landeplage committed:
refactor: Removed old audio implementation classes

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs:
```diff
@@ -62,6 +61,9 @@ public class AvatarEditorWearableFilter
     internal readonly Dictionary<string, ItemSelector> selectorsByCategory = new Dictionary<string, ItemSelector>();
 
+    [HideInInspector]
+    public System.Action<AvatarModel> onAvatarAppear;
```
```suggestion
    public event System.Action<AvatarModel> OnAvatarAppear;
```

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarEditorHUD/Scripts/AvatarEditorHUDView.cs:
```diff
@@ -229,33 +231,8 @@ public void UpdateAvatarPreview(AvatarModel avatarModel)
-
-                
+                if (onAvatarAppear != null)
+                    onAvatarAppear.Invoke(avatarModel);
```
```suggestion
                    OnAvatarAppear?.Invoke(avatarModel);
```

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/ChatWidgetHUD/ChatEntry.cs:
```diff
@@ -133,42 +133,42 @@ public void Populate(Model chatEntryModel)
             }
 
-            HUDAudioPlayer.i.RefreshChatLastCheckedTimestamp();
+            HUDAudioHandler.i.RefreshChatLastCheckedTimestamp();
```
I think we can move this `RefreshChatLastCheckedTimestamp` to this class, and remove `HUDAudioHandler` completely. 😄

Landeplage replied:
Yes! I'm a bit unclear on which of these I should use. I'll experiment and see if I can figure it out.
![image](https://user-images.githubusercontent.com/1711394/92610239-c4a05400-f2b7-11ea-897b-02f2aa660851.png)

BrianAmadori replied:
I think you can use the `lastReadWorldChatMessages` one.

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/MapRenderer/MapRenderer.cs:
```diff
@@ -74,6 +73,9 @@ public bool parcelHighlightEnabled
         private bool isInitialized = false;
 
+        [HideInInspector]
+        public System.Action onMovedParcelCursor;
```
```suggestion
        public event System.Action OnMovedParcelCursor;
```

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/MapRenderer/MapRenderer.cs:
```diff
@@ -206,9 +208,10 @@ void UpdateParcelHighlight()
             {
-                audioHandler.PlayMapParcelHighlight();
+                if (onMovedParcelCursor != null)
+                    onMovedParcelCursor.Invoke();
```
```suggestion
                    OnMovedParcelCursor?.Invoke();
```

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/Audio/HUDElements/SliderAudioHandler.cs:
```diff
@@ -4,57 +4,44 @@
+    void Awake()
     {
-        audioPlayer = HUDAudioPlayer.i;
+        eventValueChanged = Resources.Load<AudioEvent>("ScriptableObjects/AudioEvents/HUDCommon/SliderValueChange");
```
Can't we move this to `AudioScriptableObjects`?

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/Common/Audio/HUDElements/ToggleAudioHandler.cs:
```diff
@@ -4,38 +4,44 @@
-        audioPlayer = HUDAudioPlayer.i;
+        eventEnable = Resources.Load<AudioEvent>("ScriptableObjects/AudioEvents/HUDCommon/Enable");
+        eventDisable = Resources.Load<AudioEvent>("ScriptableObjects/AudioEvents/HUDCommon/Disable");
+
```
Can't we move those to `AudioScriptableObjects`?

BrianAmadori commented on unity-client/Assets/Scripts/MainScripts/DCL/Controllers/Audio/AudioContainer.cs:
```diff
@@ -38,14 +33,12 @@ void Awake()
-            // Set up AudioSource component for event
-            e.source = gameObject.AddComponent(typeof(AudioSource)) as AudioSource;
+            // Add AudioSource component for event
+            e.Initialize(gameObject.AddComponent(typeof(AudioSource)) as AudioSource);
```
Maybe we can just pass the `AudioContainer` here, and move all the `AddComponent` and `AudioSource` setup inside that method? This way the entire initialization gets encapsulated inside `AudioEvent`.

BrianAmadori reviewed (changes requested):
Orgasmic refactor. Left some minor requests to put the cherry on top of this PR. 🍒

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation

Landeplage committed:
refactor: Minor changes to audio implementation

Use event keyword for actions
Load enable/disable and slider value change audio events in common scriptable objects class.

Landeplage committed:
refactor: Moved AudioSource setup from AudioContainer to AudioEvent

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation

BrianAmadori reviewed (approved):
🚀

Landeplage committed:
Merge branch 'master' into refactor/audio-implementation
